### PR TITLE
feat: Add loadESLint() API method for v8

### PR DESCRIPTION
**Pull request by nzakas**
**Note:** Intentionally sent to v8.x branch.

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

- Added the `loadESLint()` method to the public API.
- Updated the `shouldUseFlatConfig()` function accept an optional `cwd` argument.
- Added `configType` to both `ESLint` and `FlatESLint`.
- Added tests for all of the above.
- Added docs to the Node.js API document.

refs #18075

### CODE REVIEW

File: tests/lib/api.js [hunk 1]
```diff
@@ -37,4 +42,27 @@ describe("api", () => {
...
+            assert.strictEqual(await api.loadESLint({ useFlatConfig: true }), FlatESLint);
+        });
+
+        it("should return ESLint when useFlatConfig is false", async () => {
+            assert.strictEqual(await api.loadESLint({ useFlatConfig: false }), api.ESLint);
+        });
+
+        it("should return FlatESLint when useFlatConfig is not provided because we have eslint.config.js", async () => {
+            assert.strictEqual(await api.loadESLint(), FlatESLint);
+        });
```
Comment (snitin315) [inline tests/lib/api.js:65]: Can we create a test case where `api.loadESLint()` returns legacy eslint?

Comment (nzakas) [reply tests/lib/api.js:65]: Hmm, good question. Let me look into that.

**Review by snitin315 — APPROVED** [review]
LGTM, thanks! I'll leave it open for others to review.

File: tests/lib/api.js [hunk 2]
```diff
@@ -37,4 +42,49 @@ describe("api", () => {
...
+
+        it("should return ESLint when useFlatConfig is not provided because we have .eslintrc.cjs", async () => {
+            assert.strictEqual(await api.loadESLint({
+                cwd: "./tests/fixtures/config-file/cjs"
+            }), FlatESLint);
+        });
```
Comment (mdjermanovic) [inline tests/lib/api.js:76]: According to its title, this test expects `ESLint`, but checks if `FlatESLint` was returned? Also, the presence of `.eslintrc.cjs` doesn't affect the logic.

Perhaps we want to test something like this:

```suggestion
        it("should return ESLint when useFlatConfig is not provided and there is no eslint.config.js", async () => {
            assert.strictEqual(await api.loadESLint({
                cwd: os.tmpdir()
            }), api.ESLint);
        });
```

File: tests/lib/api.js [hunk 3]
```diff
@@ -10,14 +10,19 @@
 //-----------------------------------------------------------------------------
 
 const assert = require("chai").assert,
-    api = require("../../lib/api");
+    api = require("../../lib/api"),
+    { FlatESLint } = require("../../lib/eslint");
```
Comment (mdjermanovic) [inline tests/lib/api.js:14]:
```suggestion
    { FlatESLint } = require("../../lib/eslint"),
    os = require("os");
```

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**